flink-tests/src/test/java/org/apache/flink/test/checkpointing/ChangelogPeriodicMaterializationTestBase.java

```diff
@@ -0,0 +1,485 @@
+import java.util.function.BooleanSupplier;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.apache.flink.shaded.guava30.com.google.common.collect.Iterables.get;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertSame;
+
+/** Base class for tests related to period materialization of ChangelogStateBackend. */
+@RunWith(Parameterized.class)
+public abstract class ChangelogPeriodicMaterializationTestBase extends TestLogger {
+
+    private static final int NUM_TASK_MANAGERS = 1;
+    private static final int NUM_TASK_SLOTS = 4;
+    protected static final int NUM_SLOTS = NUM_TASK_MANAGERS * NUM_TASK_SLOTS;
+    protected static final int TOTAL_ELEMENTS = 10_000;
+
+    protected final AbstractStateBackend delegatedStateBackend;
+
+    protected MiniClusterWithClientResource cluster;
+
+    // Maintain the RunnableFuture of SnapshotResult to make sure the materialization phase has
+    // been done
+    protected static final Queue<RunnableFuture<SnapshotResult<KeyedStateHandle>>>
+            PENDING_MATERIALIZATION = new ConcurrentLinkedQueue<>();
+
+    @Rule public Timeout globalTimeout = Timeout.seconds(60);
+
+    @ClassRule public static final TemporaryFolder TEMPORARY_FOLDER = new TemporaryFolder();
+
+    @Parameterized.Parameters(name = "delegated state backend type ={0}")
+    public static Collection<AbstractStateBackend> parameter() {
+        return Arrays.asList(
+                new HashMapStateBackend(),
+                new EmbeddedRocksDBStateBackend(true),
+                new EmbeddedRocksDBStateBackend());
+    }
+
+    public ChangelogPeriodicMaterializationTestBase(AbstractStateBackend delegatedStateBackend) {
+        this.delegatedStateBackend = delegatedStateBackend;
+    }
+
+    @Before
+    public void setup() throws Exception {
+        cluster =
+                new MiniClusterWithClientResource(
+                        new MiniClusterResourceConfiguration.Builder()
+                                .setConfiguration(configureDSTL())
+                                .setNumberTaskManagers(NUM_TASK_MANAGERS)
+                                .setNumberSlotsPerTaskManager(NUM_TASK_SLOTS)
+                                .build());
+        cluster.before();
+        cluster.getMiniCluster().overrideRestoreModeForChangelogStateBackend();
+
+        // init source data randomly
+        ControlledSource.initSourceData(TOTAL_ELEMENTS);
+    }
+
+    @After
+    public void tearDown() throws IOException {
+        cluster.after();
+        // clear result in sink
+        CollectionSink.clearExpectedResult();
+        PENDING_MATERIALIZATION.clear();
+    }
+
+    protected StreamGraph buildStreamGraph(
+            StreamExecutionEnvironment env, ControlledSource controlledSource) {
+        env.addSource(controlledSource)
+                .keyBy(element -> element)
+                .map(new CountMapper())
```

Comment:
I'm not sure whether modifying `CountMapper` will help, because the key can be selected for backend before processing.
Why not just add `% 100` here, in `.keyBy`?

Reply:
I have modified `ControlledSource` and it will just send `element % 100`.